### PR TITLE
refactor(data): flatten datasetsrc into flat copy_dataset_root field

### LINKED ISSUE #1502: refactor(data): flatten datasetsrc wrapper into a flat copy_dataset_root field
## Problem

The dataset-copy source is a single-field wrapper model (`DatasetSrcConfig.copy_dataset_root`) nested under `DatasetSpec.datasetsrc`, defaulting to `datasetsrc: null` in `dataset.yaml`. That null default is a Hydra struct footgun: a dotted `datasetsrc.copy_dataset_root=<path>` override fails because OmegaConf cannot index into a null node, forcing the awkward mapping form `datasetsrc='{copy_dataset_root: <path>}'`.

## Change

Replace the wrapper with a flat `DatasetSpec.copy_dataset_root: str | None` (`copy_dataset_root: null` in `dataset.yaml`), so the override is simply `copy_dataset_root=<path>`. The blank-string and hdf5-only validators move onto `DatasetSpec`. A `mode="before"` back-compat shim (mirroring `_normalize_r2_input`) promotes legacy `datasetsrc` keys so `input_spec.json` files already materialized to R2 still load.

Part of the dataset-copy feature work (#1429).

### PR DESCRIPTION
**Pull request by ktinubu**
## What

Flatten the single-field dataset-copy wrapper `DatasetSrcConfig.copy_dataset_root` (nested under `DatasetSpec.datasetsrc`) into a flat `DatasetSpec.copy_dataset_root: str | None`.

## Why

`datasetsrc` was a one-field wrapper defaulting to `datasetsrc: null` in `dataset.yaml`. That null default is a Hydra struct footgun: a dotted `datasetsrc.copy_dataset_root=<path>` override fails because OmegaConf can't index into a null node, forcing the awkward mapping form `datasetsrc='{copy_dataset_root: <path>}'`. The wrapper bought no structure (YAGNI). With the flat field the override is simply `copy_dataset_root=<path>`.

## How

- `spec.py`: replace the wrapper with `copy_dataset_root: str | None`; move the blank-string and hdf5-only validators onto `DatasetSpec`; drop `DatasetSrcConfig` (and its `__all__` export).
- **Back-compat:** a `mode="before"` shim (`_promote_legacy_datasetsrc`, mirroring `_normalize_r2_input`) promotes legacy `datasetsrc: {copy_dataset_root: X}` to the flat key and drops `datasetsrc: null`, so `input_spec.json` files already materialized to R2 still load. Verified against an existing smoke-shard spec.
- `cli/generate_dataset.py`, `configs/dataset.yaml`, and docs updated to the flat key.
- Tests cover the flat field (default/blank/hdf5-only/round-trip) and the shim edges (mapping promoted, null dropped, both-keys rejected, non-mapping rejected, empty mapping disables copy, legacy JSON reload via `model_validate_json`).

## Verification

- `make format` clean (ruff, pyright, pydoclint, prettier, mdformat, codespell).
- Pipeline test suite green (783 passed, bounded workers — `make test-fast`'s `-n auto` crashes the runner at the xdist layer, unrelated to this change).
- Pre-PR review: 7-skill `/repo-review-full-no-comments` pass — 0 BLOCK; the one gate-blocking comment-hygiene WARN and the test-coverage WARNs were addressed in the second commit.

Closes #1502
Refs #1429

### CODE REVIEW

**Comment by ktinubu** [comment]
## Verification Results

Behavioral verification of the `datasetsrc` → flat `copy_dataset_root` flatten. The audio-render path is **unchanged** by this PR (it consumes the same `--copy_dataset_root` subprocess flag), so checks target the changed surface: Hydra override ergonomics, the launch-time copy preflight, and back-compat loading of already-materialized specs. A full copy *render* would exercise only the untouched renderer, so it is intentionally not the Level 1 check here (scope-matching).

- [x] **End-to-end copy preflight works on a real materialized source** *(Level 1 — exercises the full changed launch path: flat field → `_validate_copy_source` → load source `input_spec.json` via the back-compat shim → `validate_copy_source`)*
  ```bash
  $ python -c "
  import json
  from synth_setter.pipeline.spec_io import load_spec_from_uri
  from synth_setter.pipeline.schemas.spec import DatasetSpec
  from synth_setter.cli.generate_dataset import _validate_copy_source
  src='/home/build/synth-setter/data/smoke-shard/smoke-shard-20260606T051355679Z'
  source=load_spec_from_uri(src+'/input_spec.json')
  data=json.loads(source.model_dump_json()); data['copy_dataset_root']=src
  _validate_copy_source(DatasetSpec.model_validate(data))"
  INFO  synth_setter.cli.generate_dataset:_validate_copy_source - dataset-copy source OK: .../input_spec.json matches the target spec
  ```

- [x] **The new flat `copy_dataset_root=<path>` override composes (the bug this PR fixes)** *(Level 2 — invokes the real Hydra entrypoint; promise is override resolution, not a full render)*
  ```bash
  $ python src/synth_setter/cli/generate_dataset.py \
      experiment=generate_dataset/smoke-shard-with-oracle-eval render.param_spec_name=surge_xt \
      train_val_test_sizes=[10,10,10] render.samples_per_shard=10 \
      copy_dataset_root=/home/build/synth-setter/data/smoke-shard/smoke-shard-20260606T051355679Z \
      --cfg job --resolve | grep '^copy_dataset_root:'
  copy_dataset_root: /home/build/synth-setter/data/smoke-shard/smoke-shard-20260606T051355679Z
  ```

- [x] **Legacy materialized spec (carries `datasetsrc: null`) still loads via the shim** *(Level 3 — loads a real pre-flatten `input_spec.json` artifact through `model_validate_json`)*
  ```bash
  $ python -c "
  from synth_setter.pipeline.spec_io import load_spec_from_uri
  s=load_spec_from_uri('/home/build/synth-setter/data/smoke-shard/smoke-shard-20260606T051355679Z/input_spec.json')
  print('copy_dataset_root =', s.copy_dataset_root)"
  copy_dataset_root = None
  ```

- [x] **Old dotted override fails fast — no silent misconfig** *(Level 2, adversarial — confirms the removed nested key is rejected at compose, not silently dropped)*
  ```bash
  $ python src/synth_setter/cli/generate_dataset.py experiment=generate_dataset/smoke-shard-with-oracle-eval \
      render.param_spec_name=surge_xt \
      datasetsrc.copy_dataset_root=/home/build/synth-setter/data/smoke-shard/smoke-shard-20260606T051355679Z --cfg job
  omegaconf.errors.ConfigKeyError: Key 'datasetsrc' is not in struct
  hydra.errors.ConfigCompositionException: Could not override 'datasetsrc.copy_dataset_root'.
  ```

Shim edge cases (non-mapping rejected, empty mapping disables copy, both-keys rejected, legacy-JSON reload) are pinned by `tests/pipeline/schemas/test_dataset_spec.py::TestCopyDatasetRoot` (8 tests, all green).

File: src/synth_setter/pipeline/schemas/spec.py [hunk 1]
```diff
@@ -792,6 +746,40 @@ def _normalize_r2_input(cls, data: Any) -> Any:
...
+        data = dict(data)
+        legacy = data.pop("datasetsrc")
+        if "copy_dataset_root" in data:
+            raise ValueError(
+                "DatasetSpec received both legacy 'datasetsrc' and 'copy_dataset_root'; "
+                "pass one shape, not both"
+            )
+        if isinstance(legacy, dict):
+            data["copy_dataset_root"] = legacy.get("copy_dataset_root")
+        elif legacy is not None:
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/synth_setter/pipeline/schemas/spec.py:798]: Good catch — fixed in 581cfef3. The shim now rejects any key other than `copy_dataset_root` in a legacy `datasetsrc` mapping (`ValueError: ... unexpected keys ...`), restoring the single-key `extra="forbid"` strictness the removed `DatasetSrcConfig` enforced. Covered by `test_legacy_datasetsrc_mapping_with_unexpected_key_is_rejected`.

File: src/synth_setter/pipeline/schemas/spec.py [hunk 2]
```diff
@@ -792,6 +746,40 @@ def _normalize_r2_input(cls, data: Any) -> Any:
...
+            raise ValueError(
+                f"legacy 'datasetsrc' must be a mapping or null, got {type(legacy).__name__}"
+            )
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/synth_setter/pipeline/schemas/spec.py:801]: Fixed in 581cfef3 (same finding as the earlier pass). `_promote_legacy_datasetsrc` now rejects any key other than `copy_dataset_root` in a legacy `datasetsrc` mapping with a `ValueError`, restoring the `extra="forbid"` strictness of the removed `DatasetSrcConfig`. Covered by `test_legacy_datasetsrc_mapping_with_unexpected_key_is_rejected`.

File: src/synth_setter/pipeline/schemas/spec.py [hunk 3]
```diff
@@ -792,6 +746,49 @@ def _normalize_r2_input(cls, data: Any) -> Any:
...
 
+    @model_validator(mode="before")
+    @classmethod
+    def _promote_legacy_datasetsrc(cls, data: Any) -> Any:
+        """Promote a legacy nested ``datasetsrc`` into the flat ``copy_dataset_root``.
+
+        Back-compat for ``input_spec.json`` files materialized before the
+        single-field ``datasetsrc`` wrapper was flattened: ``datasetsrc: null``
+        is dropped (the field defaults to ``None``) and
+        ``datasetsrc: {copy_dataset_root: X}`` becomes ``copy_dataset_root: X``.
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/synth_setter/pipeline/schemas/spec.py:768]: Fixed in baa414c2. `from_hydra_cfg` now raises a migration error if the composed cfg carries a stale `datasetsrc` key (`'datasetsrc' is no longer a config key ... Use copy_dataset_root=<path>`), so a Hydra-side override can no longer be silently masked into a no-op. Chose option (b) — the shim still handles materialized `input_spec.json` back-compat; Hydra overrides must use the new key. Covered by `test_stale_datasetsrc_key_is_rejected_with_migration_pointer`.

File: src/synth_setter/pipeline/schemas/spec.py [hunk 4]
```diff
@@ -792,6 +746,49 @@ def _normalize_r2_input(cls, data: Any) -> Any:
...
+            )
+        if isinstance(legacy, dict):
+            unexpected = set(legacy) - {"copy_dataset_root"}
+            if unexpected:
+                raise ValueError(
+                    f"legacy 'datasetsrc' mapping has unexpected keys {sorted(unexpected)}; "
+                    "only 'copy_dataset_root' is supported"
+                )
+            data["copy_dataset_root"] = legacy.get("copy_dataset_root")
+        elif legacy is not None:
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/synth_setter/pipeline/schemas/spec.py:798]: Fixed in baa414c2. A non-null legacy `datasetsrc` mapping must now hold exactly a non-null `copy_dataset_root` — an empty mapping or null inner value raises (`must hold exactly a non-null 'copy_dataset_root'; use datasetsrc: null to disable copy`), restoring the removed `DatasetSrcConfig`'s required-field contract. Covered by `test_legacy_datasetsrc_empty_mapping_is_rejected` and `test_legacy_datasetsrc_null_inner_is_rejected`.

**Comment by ktinubu** [comment]
**CI note — "VST slow tests (Docker)":** this check was *cancelled at the 60-min job timeout* (`test-vst-slow.yml` `timeout-minutes: 60`), not a test-assertion failure. It is a chronically slow job that hits the same timeout on `main` (3 of the last 5 `main` runs of the *VST Slow Tests* workflow were cancelled), and it is a non-required check (`mergeStateStatus=UNSTABLE`, not `BLOCKED`). This PR is a schema-only flatten of the dataset-copy config and touches no renderer/benchmark code, so the timeout is independent of the change. Re-running the job once to give it a clean shot.

**Comment by ktinubu** [comment]
**Correction:** I cannot re-trigger the timed-out "VST slow tests (Docker)" job from here — both `gh run rerun --failed` and `gh workflow run test-vst-slow.yml` return `403 Resource not accessible by personal access token`. A maintainer can re-run that single job from the Actions tab (it's a non-required check, so it does not block merge). To recap why it's safe to disregard: it was *cancelled at the 60-min job timeout*, it times out the same way on `main` (3 of the last 5 *VST Slow Tests* runs on `main` were cancelled), and this PR changes only the dataset-copy config schema — no renderer/benchmark code — so the timeout is independent of the change.

**Comment by ktinubu** [comment]
## /pr-readiness — CI note: "VST slow tests (Docker)"

This check was **cancelled at the 60-min job timeout** (`test-vst-slow.yml` `timeout-minutes: 60`, run ended at 1h0m17s), **not a test-assertion failure**.

- It is a **chronically slow job** that hits the same timeout on `main`: **3 of the last 5 `main` runs** of the *VST Slow Tests* workflow were cancelled (2026-06-06 00:28, 2026-06-05 23:46, 2026-06-05 23:10).
- It is a **non-required check** — it does not gate merge.
- This PR is a **schema-only flatten of the dataset-copy config** and touches no renderer/benchmark code, so the timeout is **independent of the change**.

**Update:** the timeout fix has been moved out of this PR into its own isolated PR — **#1522** (`ci(vst-slow): raise VST Slow Tests job timeout to 120 min`, tracking #1521). This PR no longer carries any workflow change.

The earlier `main` **merge conflict** has been resolved here by merging `origin/main` (uv.lock regenerated; schema-flatten + generate_dataset tests pass locally). `mergeable` is now `MERGEABLE`.
